platform/posix/ota_pal/source/ota_pal_posix.c

```diff
@@ -0,0 +1,626 @@
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+
+
+/* OTA PAL implementation for linux platform. */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <assert.h>
+#include <libgen.h>
+
+#include "ota.h"
+#include "ota_pal_posix.h"
+
+#include <openssl/evp.h>
+#include <openssl/bio.h>
+#include <openssl/x509.h>
+#include <openssl/pem.h>
+
+
+/* Size of buffer used in file operations on this platform (linux). */
+#define OTA_PAL_LINUX_BUF_SIZE    ( ( size_t ) 4096U )
+
+
+/* Specify the OTA signature algorithm we support on this platform. */
+const char OTA_JsonFileSignatureKey[ OTA_FILE_SIG_KEY_STR_MAX_LENGTH ] = "sig-sha256-ecdsa";
+
+static OtaPalStatus_t otaPal_CheckFileSignature( OtaFileContext_t * const C );
+
+/* Read the specified signer certificate from the filesystem into a local buffer. The allocated
+ * memory becomes the property of the caller who is responsible for freeing it. */
+static EVP_PKEY * Openssl_GetPkeyFromCertificate( uint8_t * pCertFilePath )
+{
+    BIO * pBio = NULL;
+    X509 * pCert = NULL;
+    EVP_PKEY * pPkey = NULL;
+    int32_t rc = 0;
+
+    /* Read the cert file */
+    pBio = BIO_new( BIO_s_file() );
+
+    if( pBio != NULL )
+    {
+        /* The operand "1" is not exposed and controllable here */
+        /* coverity[misra_c_2012_rule_10_1_violation] */
+        rc = BIO_read_filename( pBio, pCertFilePath );
+
+        if( rc != 1 )
+        {
+            LogDebug( ( " TEMP solution: No cert file, get the signer cert from a pre-defined variable\n" ) );
+
+            /* Get the signer cert from a predefined PEM string */
+            BIO_free_all( pBio );
+            pBio = BIO_new( BIO_s_mem() );
+
+            if( pBio != NULL )
+            {
+                rc = BIO_puts( pBio, signingcredentialSIGNING_CERTIFICATE_PEM );
+
+                if( rc <= 0 )
+                {
+                    LogError( ( "Failed to write a PEM string to BIO stream" ) );
+                }
+            }
+            else
+            {
+                LogError( ( "Failed to read certificate from a PEM string." ) );
+            }
+        }
+        else
+        {
+            LogError( ( "Failed to read certificate from a file." ) );
+        }
+    }
+
+    if( pBio != NULL )
+    {
+        pCert = PEM_read_bio_X509( pBio, NULL, NULL, NULL );
+
+        if( pCert != NULL )
+        {
+            LogDebug( ( "Getting the pkey from the X509 cert." ) );
+
+            /* Extract the public key */
+            pPkey = X509_get_pubkey( pCert );
+
+            if( pPkey == NULL )
+            {
+                LogError( ( "Failed to get pkey from the signer cert." ) );
+            }
+        }
+        else
+        {
+            LogError( ( "Failed to load cert from either file or predefined string." ) );
+        }
+    }
+    else
+    {
+        LogError( ( "Failed to read signer cert." ) );
+    }
+
+    BIO_free_all( pBio );
```

Comment:
Is it okay to call these when pBio and pCert are NULL?

Reply:
yes, BIO_free_all will take care of NULL pointers